recommenders/models/sar/sar_singlenode.py

```diff
@@ -192,27 +203,32 @@ def set_index(self, df):
 
         # generate a map of continuous index values to items
         self.index2item = dict(enumerate(df[self.col_item].unique()))
+        self.index2user = dict(enumerate(df[self.col_user].unique()))
 
-        # invert the mapping from above
+        # invert the mappings from above
         self.item2index = {v: k for k, v in self.index2item.items()}
-
-        # create mapping of users to continuous indices
-        self.user2index = {x[1]: x[0] for x in enumerate(df[self.col_user].unique())}
+        self.user2index = {v: k for k, v in self.index2user.items()}
 
         # set values for the total count of users and items
         self.n_users = len(self.user2index)
         self.n_items = len(self.index2item)
 
-    def fit(self, df):
+    def fit(self, df, compute_user_similarity=False):
         """Main fit method for SAR.
 
         .. note::
 
-        Please make sure that `df` has no duplicates.
+        Please make sure that `df` has no duplicate user-item pairs.
 
         Args:
-            df (pandas.DataFrame): User item rating dataframe (without duplicates).
+            df (pandas.DataFrame): User item rating dataframe (without duplicate user-items).
+            compute_user_similarity (bool): If true, compute user similarity using the self.similarity_type. Item
+            similarity is always calculated. This must be true to use self.get_user_based_topk to get similar users.
         """
+        if sum(df[[self.col_user, self.col_item]].duplicated()) > 0:
```

Comment:
I think either option is better than just having a docstring. If there is a use case for running SAR with duplicates, then I think warning is best. But, if we believe that the algorithm should NEVER see duplicates, then I would choose error. 

If we do choose to go with error, we will need to fix multiple tests. If we go with warning, we don't need to fix the tests (but still probably should IMO). I was just hesitant to change a bunch of tests without first having this discussion. 

Let me know what you all think.

Reply:
So currently, yes duplicates will give us weird results in the cooccurrence matrix. We can fix this by simply removing duplicates from a copy of the dataframe before calculating cooccurrence matrix like this:

Reply:
@danb27 for this I would say that it is better to check whether there are duplicates, and if so, throw an error.
In my experience `df.duplicated().any()` is faster than `any(df.duplicated())` but not sure if your implementation is faster. In `sum(df[[self.col_user, self.col_item]].duplicated())` I would consider also ratings column. I can see a case where the ratings  are related to clicks and purchase, where both signals are valid, so we want to keep both.

Reply:
![image](https://user-images.githubusercontent.com/42096328/161596333-2166ad41-34d3-4576-a7ed-fe104e17c8fe.png)
@miguelgfierro You are correct, I shouldn't be using sum because then it needs to add up all the True/False in the array resulting from df.duplicated(). Using any, we only need to check until there is a single True - so our best case is a lot better. 

The bigger issue here is that if I implement this error like this (to accommodate your comments and @anargyri 's comments):
```
        select_columns = [self.col_user, self.col_item, self.col_rating]
        if self.time_decay_flag:
            select_columns += [self.col_timestamp]

        if df[select_columns].duplicated().any():
            raise ValueError("There should not be duplicates in the dataframe")
```
then four tests will fail: test_user_affinity, test_recommend_k_items[3-cooccurrence-count], test_recommend_k_items[3-jaccard-jac], test_recommend_k_items[3-lift-lift]

It seems odd to change these tests, since the dataset being used [https://github.com/microsoft/recommenders/blob/main/tests/conftest.py#L91](url) was designed for SAR. I don't know enough about this dataset to know what to do here. 

What do you think?

Reply:
@simonzhaoms please see Dan's message above. It looks some of the tests we have both in SAR CPU and SAR+ rely on a dataset with duplicates. 

@danb27 can you tell me exactly which csv have duplicate values? I can create a non duplicate version and upload it to the blob.

FYI @anargyri

Reply:
@miguelgfierro The csv file with duplicates is located at  https://recodatasets.z20.web.core.windows.net/sarunittest/demoUsage.csv 

I tested locally after deduping with this code: 

```
df = df.sort_values(
    "Timestamp", ascending=False
)
df = df.drop_duplicates(
    ["UserId", "MovieId", "Timestamp"],
    keep="first"
)
```

which is enough to pass all of the unit tests with the code I just pushed. The tests will fail for now, but once you get the deduped version up in storage I can swap the link and everything should be ready to go.

Reply:
I'm getting this error: `urllib.error.HTTPError: HTTP Error 404: The requested content does not exist.` with the above link

Reply:
Sorry, I didn't set the right permissions. Now it should work

Reply:
tests passed locally! PR is hopefully good to go now after latest commit.